src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs

```diff
@@ -5349,5 +5349,303 @@ static void Main()
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceUserStruct", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""Program.S Program.S?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceUserStruct", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""Program.S Program.S?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+            comp.VerifyIL("Program.CoalesceStructWithImplicitConstructor", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""Program.S Program.S?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCoalesce_NullableWithConvertedDefault_Optimization()
+        {
+            var source = @"
+class Program
+{
+    static (bool a, System.Guid b, string c) CoalesceDifferentTupleNames((bool a, System.Guid b, string c)? x)
+    {
+        return x ?? default((bool c, System.Guid d, string e));
+    }
+
+    static void Main()
+    {
+        System.Console.WriteLine(CoalesceDifferentTupleNames((true, new System.Guid(""533d4d3b-5013-461e-ae9e-b98eb593d761""), ""foo"")));
+        System.Console.WriteLine(CoalesceDifferentTupleNames(null));
+    }
+}";
+            var expectedOutput = 
+ @"(True, 533d4d3b-5013-461e-ae9e-b98eb593d761, foo)
+(False, 00000000-0000-0000-0000-000000000000, )";
+            var comp = CompileAndVerify(source, expectedOutput: expectedOutput);
+            comp.VerifyIL("Program.CoalesceDifferentTupleNames", @"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""(bool a, System.Guid b, string c) (bool a, System.Guid b, string c)?.GetValueOrDefault()""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNullCoalesce_NullableWithNonDefault_NoOptimization()
+        {
+            var source = @"
+class Program
+{
+    struct S
+    {
+    }
+
+    static int? CoalesceWithNonDefault1(int? x)
+    {
+        return x ?? 2;
+    }
+
+    static int? CoalesceWithNonDefault2(int? x, int? y)
+    {
+        return x ?? y;
+    }
+
+    static int? CoalesceWithNonDefault3(int? x)
+    {
+        return x ?? default(int?);
+    }
+
+    static void Main()
+    {
+        System.Console.WriteLine(CoalesceWithNonDefault1(42));
+        System.Console.WriteLine(CoalesceWithNonDefault1(null));
+        System.Console.WriteLine(CoalesceWithNonDefault2(123, 456));
+        System.Console.WriteLine(CoalesceWithNonDefault2(123, null));
+        System.Console.WriteLine(CoalesceWithNonDefault2(null, 456));
+        System.Console.WriteLine(CoalesceWithNonDefault2(null, null));
+        System.Console.WriteLine(CoalesceWithNonDefault3(42));
+        System.Console.WriteLine(CoalesceWithNonDefault3(null));
```

Comment:
>CoalesceWithNonDefault3(null) [](start = 33, length = 29)

For cases that expected to return null, it is better to output something recognizable so that it could be easily found in the base line. For example, it looks like the output for this line is simply missing from the base line and the test still passes. The reader is likely to be confused as well. This could be done like this, for example:
```
System.Console.WriteLine(CoalesceWithNonDefault3(null)?.ToString() ?? "*null*");
```

Reply:
Done